### PR TITLE
allow changes to the array base column type in migrations

### PR DESCRIPTION
**Pull request by sinisaos**
Related to #1076

### CODE REVIEW

File: piccolo/apps/migrations/auto/migration_manager.py [hunk 1]
```diff
@@ -479,6 +479,35 @@ async def _run_alter_columns(self, backwards: bool = False):
...
+                            _Table.alter().drop_column(old_array_column)
+                        )
+                        # than add array column with new base column type
+                        await self._run_query(
+                            _Table.alter().add_column(
+                                name=new_array_column._meta.name,
+                                column=new_array_column,
+                            )
+                        )
                     if old_column_class != column_class:
```
Comment (dantownsend) [inline piccolo/apps/migrations/auto/migration_manager.py:511]: Rather than the logic you've added above, can't you just change this `if` statement to something like:

```python
base_colum_class = params.get('base_column').__class__
old_base_column_class = old_params.get('base_column').__class__

if (old_column_class != column_class) or (old_base_colum_class != base_colum_class):
    ...
```

File: piccolo/apps/migrations/auto/migration_manager.py [hunk 2]
```diff
@@ -479,6 +479,35 @@ async def _run_alter_columns(self, backwards: bool = False):
...
+                        )
+
+                        new_array_column = column_class(**params)
+                        new_array_column._meta._table = _Table
+                        new_array_column._meta._name = alter_column.column_name
+                        new_array_column._meta.db_column_name = (
+                            alter_column.db_column_name
+                        )
+                        # first drop array column
+                        await self._run_query(
```
Comment (dantownsend) [inline piccolo/apps/migrations/auto/migration_manager.py:501]: We can't drop the column and recreate it because there might be data in it.

**Review by dantownsend — COMMENTED** [review]
Thanks for trying to fix this - I would really like to get it done.

File: piccolo/apps/migrations/auto/migration_manager.py (hunk 1 above)
Comment (sinisaos) [reply piccolo/apps/migrations/auto/migration_manager.py:511]: Great. Thanks. That's much better.

**Review by Skelmis — CHANGES_REQUESTED** [review]
This pull request has not worked for my test case seen below

1. Table
```py
from piccolo.columns import Array, Text, Integer
from piccolo.table import Table

class TestTable(Table):
    my_col = Array(base_column=Integer())
```

2. Create and run migration
3. Modify table to
```py
from piccolo.columns import Array, Text, Integer
from piccolo.table import Table

class TestTable(Table):
    my_col = Array(base_column=Text())
```
4. Create and run migrations resulting in the following traceback

```text
near "ALTER": syntax error
Traceback (most recent call last):
  File "/home/skelmis/tmp/piccolo_array_pr/venv/lib/python3.12/site-packages/targ/__init__.py", line 449, in run
    command.call_with(arg_class)
  File "/home/skelmis/tmp/piccolo_array_pr/venv/lib/python3.12/site-packages/targ/__init__.py", line 230, in call_with
    asyncio.run(self.command(**cleaned_kwargs))
  File "/usr/lib/python3.12/asyncio/runners.py", line 195, in run
    return runner.run(main)
           ^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.12/asyncio/runners.py", line 118, in run
    return self._loop.run_until_complete(task)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "uvloop/loop.pyx", line 1518, in uvloop.loop.Loop.run_until_complete
  File "/home/skelmis/tmp/piccolo_array_pr/piccolo/apps/migrations/commands/forwards.py", line 161, in forwards
    response = await run_forwards(
               ^^^^^^^^^^^^^^^^^^^
  File "/home/skelmis/tmp/piccolo_array_pr/piccolo/apps/migrations/commands/forwards.py", line 122, in run_forwards
    response = await manager.run()
               ^^^^^^^^^^^^^^^^^^^
  File "/home/skelmis/tmp/piccolo_array_pr/piccolo/apps/migrations/commands/forwards.py", line 99, in run
    return await self.run_migrations(app_config)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/skelmis/tmp/piccolo_array_pr/piccolo/apps/migrations/commands/forwards.py", line 83, in run_migrations
    await response.run()
  File "/home/skelmis/tmp/piccolo_array_pr/piccolo/apps/migrations/auto/migration_manager.py", line 1048, in run
    await self._run_alter_columns(backwards=backwards)
  File "/home/skelmis/tmp/piccolo_array_pr/piccolo/apps/migrations/auto/migration_manager.py", line 505, in _run_alter_columns
    await self._run_query(
  File "/home/skelmis/tmp/piccolo_array_pr/piccolo/apps/migrations/auto/migration_manager.py", line 426, in _run_query
    await query.run()
  File "/home/skelmis/tmp/piccolo_array_pr/piccolo/query/base.py", line 440, in run
    return await engine.run_ddl(self.ddl[0], in_pool=in_pool)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/skelmis/tmp/piccolo_array_pr/piccolo/engine/sqlite.py", line 820, in run_ddl
    response = await self._run_in_existing_connection(
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/skelmis/tmp/piccolo_array_pr/piccolo/engine/sqlite.py", line 756, in _run_in_existing_connection
    async with connection.execute(query, args) as cursor:
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/skelmis/tmp/piccolo_array_pr/venv/lib/python3.12/site-packages/aiosqlite/context.py", line 41, in __aenter__
    self._obj = await self._coro
                ^^^^^^^^^^^^^^^^
  File "/home/skelmis/tmp/piccolo_array_pr/venv/lib/python3.12/site-packages/aiosqlite/core.py", line 183, in execute
    cursor = await self._execute(self._conn.execute, sql, parameters)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/skelmis/tmp/piccolo_array_pr/venv/lib/python3.12/site-packages/aiosqlite/core.py", line 122, in _execute
    return await future
           ^^^^^^^^^^^^
  File "/home/skelmis/tmp/piccolo_array_pr/venv/lib/python3.12/site-packages/aiosqlite/core.py", line 105, in run
    result = function()
             ^^^^^^^^^^
sqlite3.OperationalError: near "ALTER": syntax error
```

Configured with `DB = SQLiteEngine()`. I did not yet test with postgres or regular migrations

**Comment by sinisaos** [comment]
@Skelmis Thanks for your effort but [automatic migrations aren't fully supported for Sqlite because Sqlite has limited support for ALTER table](https://piccolo-orm.readthedocs.io/en/latest/piccolo/migrations/create.html#auto-migrations). With Postgres and Cockroach works. Feel free to try this PR with Postgres or Cockroach. Thanks again.

**Comment by sinisaos** [comment]
@Skelmis No problem. Thanks for the comment.

**Review by Skelmis — COMMENTED** [review]
Trying to go from `Integer` to `Serial` yields a stack trace when I'd expected it to instead say it was unsupported. I'm unsure if this stack trace is a blocking issue tho

Expected: `asyncpg.exceptions.FeatureNotSupportedError: array of serial is not implemented`
Actual:
```
`_table` isn't defined - the Table Metaclass should set it.
Traceback (most recent call last):
  File "/home/skelmis/tmp/piccolo_array_pr/venv/lib/python3.12/site-packages/targ/__init__.py", line 449, in run
    command.call_with(arg_class)
  File "/home/skelmis/tmp/piccolo_array_pr/venv/lib/python3.12/site-packages/targ/__init__.py", line 230, in call_with
    asyncio.run(self.command(**cleaned_kwargs))
  File "/usr/lib/python3.12/asyncio/runners.py", line 195, in run
    return runner.run(main)
           ^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.12/asyncio/runners.py", line 118, in run
    return self._loop.run_until_complete(task)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "uvloop/loop.pyx", line 1518, in uvloop.loop.Loop.run_until_complete
  File "/home/skelmis/tmp/piccolo_array_pr/piccolo/apps/migrations/commands/forwards.py", line 161, in forwards
    response = await run_forwards(
               ^^^^^^^^^^^^^^^^^^^
  File "/home/skelmis/tmp/piccolo_array_pr/piccolo/apps/migrations/commands/forwards.py", line 122, in run_forwards
    response = await manager.run()
               ^^^^^^^^^^^^^^^^^^^
  File "/home/skelmis/tmp/piccolo_array_pr/piccolo/apps/migrations/commands/forwards.py", line 99, in run
    return await self.run_migrations(app_config)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/skelmis/tmp/piccolo_array_pr/piccolo/apps/migrations/commands/forwards.py", line 83, in run_migrations
    await response.run()
  File "/home/skelmis/tmp/piccolo_array_pr/piccolo/apps/migrations/auto/migration_manager.py", line 1048, in run
    await self._run_alter_columns(backwards=backwards)
  File "/home/skelmis/tmp/piccolo_array_pr/piccolo/apps/migrations/auto/migration_manager.py", line 511, in _run_alter_columns
    new_column.column_type,
    ^^^^^^^^^^^^^^^^^^^^^^
  File "/home/skelmis/tmp/piccolo_array_pr/piccolo/columns/column_types.py", line 2638, in column_type
    return f"{self.base_column.column_type}[]"
              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/skelmis/tmp/piccolo_array_pr/piccolo/columns/column_types.py", line 778, in column_type
    engine_type = self._meta.engine_type
                  ^^^^^^^^^^^^^^^^^^^^^^
  File "/home/skelmis/tmp/piccolo_array_pr/piccolo/columns/base.py", line 227, in engine_type
    engine = self.table._meta.db
             ^^^^^^^^^^
  File "/home/skelmis/tmp/piccolo_array_pr/piccolo/columns/base.py", line 209, in table
    raise ValueError(
ValueError: `_table` isn't defined - the Table Metaclass should set it.
```

But going from `Text` to `Integer` works as expected :)

**Comment by Skelmis** [comment]
Yea, I was more so curious if we should deal with the `ValueError: `_table` isn't defined - the Table Metaclass should set it.` error or leave as is. The not implemented existing error is fine

**Review by Skelmis — APPROVED** [review]
Some weird side case errors that are unhandled but as I see it there is nothing blocking this merge

File: piccolo/apps/migrations/auto/migration_manager.py [hunk 3]
```diff
@@ -472,7 +472,14 @@ async def _run_alter_columns(self, backwards: bool = False):
                 if (old_column_class is not None) and (
                     column_class is not None
                 ):
-                    if old_column_class != column_class:
+                    base_colum_class = params.get("base_column").__class__
```
Comment (dantownsend) [inline piccolo/apps/migrations/auto/migration_manager.py:475]: There's a typo here - `base_colum_class` -> `base_column_class`.

File: piccolo/apps/migrations/auto/diffable_table.py [hunk 4]
```diff
@@ -188,7 +189,7 @@ def __sub__(self, value: DiffableTable) -> TableDelta:
                         tablename=self.tablename,
                         column_name=column._meta.name,
                         db_column_name=column._meta.db_column_name,
-                        params=deserialise_params(delta),
+                        params=params,
```
Comment (dantownsend) [inline piccolo/apps/migrations/auto/diffable_table.py:192]: I don't fully understand the change here - why doesn't `deserialise_params` work?

File: piccolo/apps/migrations/auto/serialisation.py [hunk 5]
```diff
@@ -521,12 +521,15 @@ def serialise_params(params: dict[str, Any]) -> SerialisedParams:
                 params=column._meta.params
             )
 
+            column_class_name = column.__class__.__name__
+            # we need to manually re-assign the base column
+            column._meta.name = f"{column_class_name.upper()}"
```
Comment (dantownsend) [inline piccolo/apps/migrations/auto/serialisation.py:526]: There's no need for an f-string here.

This works fine:

```python
column._meta.name = column_class_name.upper()
```

Why do we have to set the column's name?

File: tests/apps/migrations/auto/integration/test_migrations.py [hunk 6]
```diff
@@ -632,65 +635,6 @@ def test_decimal_column(self):
             ),
         )
 
-    @engines_skip("cockroach")
-    def test_array_column_integer(self):
```
Comment (dantownsend) [inline tests/apps/migrations/auto/integration/test_migrations.py:636]: Can we keep these tests somehow?

File: piccolo/apps/migrations/auto/migration_manager.py (hunk 3 above)
Comment (sinisaos) [reply piccolo/apps/migrations/auto/migration_manager.py:475]: Thanks. I will change that

File: piccolo/apps/migrations/auto/serialisation.py (hunk 5 above)
Comment (sinisaos) [reply piccolo/apps/migrations/auto/serialisation.py:526]: I'm resetting `base_column` here because if we don't, `ValueError: `_name` is not defined - Table Metaclass should set it` is raised when we run the migration. If you have a better solution, please change this.

File: tests/apps/migrations/auto/integration/test_migrations.py (hunk 6 above)
Comment (sinisaos) [reply tests/apps/migrations/auto/integration/test_migrations.py]: Yes, we can. I don't remember why I rewrote those tests because now they work. I'll get those tests back.

**Comment by sinisaos** [comment]
@dantownsend Thanks for the review. I've made changes and implemented your suggestions and tried to explain (in the comments) what I mean by those changes.

File: piccolo/apps/migrations/auto/diffable_table.py (hunk 4 above)
Comment (dantownsend) [reply piccolo/apps/migrations/auto/diffable_table.py:192]: OK - thanks. I remember the problem with auto migrations not importing the array's base column. I need to figure out what's going on there.

File: tests/apps/migrations/auto/integration/test_migrations.py [hunk 7]
```diff
@@ -727,7 +727,14 @@ def test_array_base_column_change(self):
                     Array(base_column=Varchar()),
                     Array(base_column=Text()),
                 ]
-            ]
+            ],
+            test_function=lambda x: all(
+                [
+                    x.data_type == "ARRAY",
+                    x.is_nullable == "NO",
+                    x.column_default == "'{}'::text[]",
```
[the remark this answers is not in the document]
Comment (dantownsend) [reply tests/apps/migrations/auto/integration/test_migrations.py:735]: @sinisaos Which version of CockroachDB are you running locally?

It's still failing for me locally, even if I expect `column_default == 'ARRAY[]'`.

[the remark this answers is not in the document]
Comment (sinisaos) [reply tests/apps/migrations/auto/integration/test_migrations.py:736]: And it fails.

[the remark this answers is not in the document]
Comment (dantownsend) [reply tests/apps/migrations/auto/integration/test_migrations.py:736]: I'm installing `25.2.2` locally - will see if that works.

[the remark this answers is not in the document]
Comment (dantownsend) [reply tests/apps/migrations/auto/integration/test_migrations.py:736]: It still fails for me. I keep on getting this error:

```
asyncpg.exceptions.InvalidTableDefinitionError: relation "my_table" (107): computed column "my_column_1" cannot also have a DEFAULT expression
```

[the remark this answers is not in the document]
Comment (dantownsend) [reply tests/apps/migrations/auto/integration/test_migrations.py:736]: Interesting - thanks!

I'm happy with upgrading our CI to use this version of CockroachDB because `v24.3` seems to be their LTS version.

[the remark this answers is not in the document]
Comment (sinisaos) [reply tests/apps/migrations/auto/integration/test_migrations.py:736]: I agree and I think you can merge this PR. If there is a problem with that test in the future, we can always skip it because only that test does not work in other versions. The code works without any problems in all versions if we actually use it in some application.

[the remark this answers is not in the document]
Comment (dantownsend) [reply tests/apps/migrations/auto/integration/test_migrations.py:736]: I think it's close. I want to test it a bit though - there may be edge cases with converting from certain array types to other array types. Once we can use the playground with Cockroach, that will help testing.

**Comment by sinisaos** [comment]
I'm closing this PR just to clean up my open requests in the repo. Since it has been open for quite a while, I'll remove it from the list to keep things tidy. Feel free to use it if needed. Thanks.
